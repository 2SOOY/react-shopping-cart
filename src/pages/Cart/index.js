import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Button from '../../components/shared/Button';
import HighlightText from '../../components/shared/HighlightText';
import NumericInput from '../../components/shared/NumericInput';
import Product from '../../components/shared/Product';
import { COLOR } from '../../constants';
import { setAllCartItemCheckbox, toggleCartItemCheckbox } from '../../store';
import {
  Container,
  Header,
  Contents,
  ListOptionMenu,
  ProductListContainer,
  ProductListWrapper,
  ProductListHeader,
  ProductList,
  ProductWrapper,
  ReceiptWrapper,
  ReceiptHeader,
  ReceiptContent,
  ReceiptRow,
  CheckBox,
} from './style';

const Cart = () => {
  const list = useSelector(state => state.cart);
  const dispatch = useDispatch();

  const checkedCount = list.filter(item => item.checked).length;
  const isAllChecked = checkedCount === list.length;
  const checkOptionText = isAllChecked
    ? '선택해제'
    : checkedCount
    ? `${checkedCount}개 선택`
    : '전체선택';

  const onCheckBoxChange = ({ id }) => {
    dispatch(toggleCartItemCheckbox(id));
  };

  const onCheckOptionChange = () => {
    dispatch(setAllCartItemCheckbox(isAllChecked));
  };

  return (
    <Container>
      <Header>장바구니</Header>
      <Contents>
        <ProductListContainer>
          <ListOptionMenu>
            <CheckBox>
              <input type="checkbox" checked={isAllChecked} onChange={onCheckOptionChange} hidden />
              <span></span>
              {checkOptionText}
            </CheckBox>
            <Button
              type="button"
              width="118px"
              height="50px"
              backgroundColor={COLOR.WHITE}
              borderColor={COLOR['GRAY-300']}
              fontSize="1rem"
            >
              상품 삭제
            </Button>
          </ListOptionMenu>
          <ProductListWrapper>
            <ProductListHeader>든든배송상품</ProductListHeader>
            <ProductList>
              {list.map(({ id, name, image, price, quantity, checked }) => (
                <ProductWrapper key={id}>
                  <CheckBox>
                    <input
                      type="checkbox"
                      onChange={() => onCheckBoxChange({ id })}
                      checked={checked}
                      hidden
                    />
                    <span></span>
                  </CheckBox>
                  <Product
                    thumbnail={{ image: image, alt: name, size: 'small' }}
                    information={{ title: name }}
                    extra={
                      <>
                        <button>휴지통</button>
                        {/* TODO: setValue 만들기 */}
                        <NumericInput value={quantity} />
                        <div>{(price * quantity).toLocaleString('ko-KR')} 원</div>
                      </>
                    }
                  />
                </ProductWrapper>
              ))}
            </ProductList>
          </ProductListWrapper>
        </ProductListContainer>
        <ReceiptWrapper>
          <ReceiptHeader>결제예상금액 헤더</ReceiptHeader>
          <ReceiptContent>
            <ReceiptRow>
              <HighlightText color={COLOR.HIGHLIGHT_MINT} fontSize="1.25rem">
                결제예상금액
              </HighlightText>
              <HighlightText color={COLOR.HIGHLIGHT_MINT} fontSize="1.25rem">
                10,000원
              </HighlightText>
            </ReceiptRow>
            <Button
              type="button"
              width="100%"
              height="74px"
              backgroundColor={COLOR.MINT}
              color={COLOR.WHITE}
              fontSize="1.5rem"
            >
              주문하기
            </Button>
          </ReceiptContent>
        </ReceiptWrapper>
      </Contents>
    </Container>
  );
};

export default Cart;
