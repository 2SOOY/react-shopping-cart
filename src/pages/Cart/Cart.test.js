import React from 'react';
import Cart from './';
import { render, fireEvent, waitFor, within } from '../../test-utils';
import '@testing-library/jest-dom/extend-expect';

const initialState = {
  cartReducer: {
    cart: [
      {
        id: 1,
        name: '[든든] 유부 슬라이스 500g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11153-main-01.png',
        price: 4900,
        quantity: 4,
        checked: true,
      },
      {
        id: 2,
        name: '[든든] 진맛살 1kg',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11157-main-01.png',
        price: 7100,
        quantity: 1,
        checked: true,
      },
    ],
  },
  reducer: {
    itemList: [],
    orderList: [],
  },
};

describe('<Cart />', () => {
  it('matches snapshot', () => {
    const utils = render(<Cart />, { initialState });

    expect(utils.container).toMatchSnapshot();
  });

  it('감소 버튼을 누르면 수가 1씩 감소하고, 해당 상품의 가격과 총 금액에 1개 만큼의 금액이 감소하고, 총 금액이 변화한다.', () => {
    const utils = render(<Cart />, { initialState });
    const number = utils.getByRole('spinbutton', {
      name: /\[든든\] 유부 슬라이스 500g 수량 변경 입력/i,
    });
    const subButton = utils.getByRole('button', {
      name: /\[든든\] 유부 슬라이스 500g 수량 변경 1개 감소/i,
    });
    const price = utils.getByLabelText('[든든] 유부 슬라이스 500g 합산 가격');
    const totalPrice = utils.getByLabelText('결제예상금액');

    fireEvent.click(subButton);
    expect(number.value).toBe('3');
    expect(price).toHaveTextContent('14,700 원');
    expect(totalPrice).toHaveTextContent('21,800원');
  });

  it('증가 버튼을 누르면 수가 1씩 증가하고, 해당 상품의 가격과 총 금액에 1개 만큼의 금액이 증가하고, 총 금액이 변화한다.', () => {
    const utils = render(<Cart />, { initialState });
    const number = utils.getByRole('spinbutton', {
      name: /\[든든\] 유부 슬라이스 500g 수량 변경 입력/i,
    });
    const addButton = utils.getByRole('button', {
      name: /\[든든\] 유부 슬라이스 500g 수량 변경 1개 증가/i,
    });
    const price = utils.getByLabelText('[든든] 유부 슬라이스 500g 합산 가격');
    const totalPrice = utils.getByLabelText('결제예상금액');

    fireEvent.click(addButton);
    expect(number.value).toBe('5');
    expect(price).toHaveTextContent('24,500 원');
    expect(totalPrice).toHaveTextContent('31,600원');
  });

  it('수량 입력 칸에 20을 입력할 시, 수량이 20이 되고, 해당 상품 가격과 총 금액이 변화한다.', () => {
    const utils = render(<Cart />, { initialState });
    const number = utils.getByRole('spinbutton', {
      name: /\[든든\] 유부 슬라이스 500g 수량 변경 입력/i,
    });
    const price = utils.getByLabelText('[든든] 유부 슬라이스 500g 합산 가격');
    const totalPrice = utils.getByLabelText('결제예상금액');

    fireEvent.change(number, { target: { value: 20 } });
    expect(number.value).toBe('20');
    expect(price).toHaveTextContent('98,000 원');
    expect(totalPrice).toHaveTextContent('105,100원');
  });

  it('장바구니 삭제 버튼 클릭 시, 해당 아이템이 삭제된다.', async () => {
    const utils = render(<Cart />, { initialState });
    const deleteButton = utils.getByRole('button', {
      name: /\[든든\] 유부 슬라이스 500g 삭제/i,
    });
    const totalPrice = utils.getByLabelText('결제예상금액');
    const confirmSpy = jest.spyOn(window, 'confirm');
    confirmSpy.mockImplementation(jest.fn(() => true));

    expect(totalPrice).toHaveTextContent('26,700원');
    fireEvent.click(deleteButton);
    expect(confirmSpy).toBeCalled();

    await waitFor(() => expect(totalPrice).toHaveTextContent('7,100원'));
  });

  it('장바구니 상품에 있는 모든 상품의 checked 상태가 true일 때 선택 해제 라벨이 보여진다.', () => {
    const utils = render(<Cart />, { initialState });
    const listOptionMenu = utils.getByLabelText('상품선택 옵션 메뉴');
    const checkOptionLabel = within(listOptionMenu).getByRole('checkbox');

    expect(checkOptionLabel).toHaveTextContent('선택해제');
    expect(checkOptionLabel).toBeChecked();
  });
});
