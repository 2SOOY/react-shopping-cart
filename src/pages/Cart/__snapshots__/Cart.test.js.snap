// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> matches snapshot 1`] = `
<div>
  <div
    class="sc-ksdxAp cFwGOh"
  >
    <h2
      class="sc-hBURRC eBjsOr"
    >
      장바구니
    </h2>
    <div
      class="sc-fotPbf dYrQej"
    >
      <div
        class="sc-fFehDp ctsaqC"
      >
        <div
          aria-label="상품선택 옵션 메뉴"
          class="sc-bkkfTU kAcZej"
        >
          <label
            aria-checked="true"
            class="sc-kfPtPH juLErD"
            role="checkbox"
          >
            <input
              checked=""
              hidden=""
              type="checkbox"
            />
            <span />
            선택해제
          </label>
          <button
            class="sc-bdvvaa VaouB"
            color="#333333"
            font-size="1rem"
            height="50px"
            type="button"
            width="118px"
          >
            상품 삭제
          </button>
        </div>
        <div
          aria-label="장바구니 상품 목록"
          class="sc-ieebsP"
        >
          <h3
            class="sc-dJjZJu cNyfZz"
          >
            배송상품 (
            2
            개)
          </h3>
          <div
            class="sc-hGPAah"
          >
            <div
              class="sc-dlVyqM dsYsil"
            >
              <label
                class="sc-kfPtPH juLErD"
              >
                <input
                  checked=""
                  hidden=""
                  type="checkbox"
                />
                <span
                  aria-checked="true"
                  aria-label="[든든] 유부 슬라이스 500g 선택"
                  role="checkbox"
                />
              </label>
              <div
                class="sc-jrQzUz ffqCKT"
              >
                <div
                  class="sc-iCfLBT"
                >
                  <div
                    class="sc-furvIG bEiCtt"
                  >
                    <img
                      alt="[든든] 유부 슬라이스 500g"
                      class="sc-pVTma gOPKiY"
                      src="https://cdn-mart.baemin.com/goods/custom/20200525/11153-main-01.png"
                    />
                  </div>
                </div>
                <div
                  class="sc-kDThTU bQjZIX"
                >
                  <h4
                    class="sc-iqsfdx hvMujD"
                  >
                    [든든] 유부 슬라이스 500g
                  </h4>
                  <div
                    class="sc-crHlIS cPCMEP"
                  />
                </div>
                <div
                  class="sc-egiSv ggtrWD"
                >
                  <button
                    aria-label="[든든] 유부 슬라이스 500g 삭제"
                    class="sc-bqiQRQ suzEO"
                    type="button"
                  >
                    <svg>
                      trash-bin.svg
                    </svg>
                  </button>
                  <div
                    class="sc-dkPtyc ibPmRP"
                  >
                    <input
                      aria-label="[든든] 유부 슬라이스 500g 수량 변경 입력"
                      class="sc-hKwCoD kXeTvV"
                      max="99"
                      min="1"
                      type="number"
                      value="4"
                    />
                    <div
                      class="sc-eCImvq isEnoV"
                    >
                      <button
                        aria-label="[든든] 유부 슬라이스 500g 수량 변경 1개 증가"
                        class="sc-jRQAMF crVgkL"
                      >
                        <img
                          alt="증가"
                          class="sc-gKckTs fNcpol"
                          src="arrow-up.svg"
                        />
                      </button>
                      <button
                        class="sc-jRQAMF crVgkL"
                      >
                        <img
                          alt="감소"
                          aria-label="[든든] 유부 슬라이스 500g 수량 변경 1개 감소"
                          class="sc-gKckTs fNcpol"
                          src="arrow-down.svg"
                        />
                      </button>
                    </div>
                  </div>
                  <div
                    aria-label="[든든] 유부 슬라이스 500g 합산 가격"
                  >
                    19,600
                     원
                  </div>
                </div>
              </div>
            </div>
            <div
              class="sc-dlVyqM dsYsil"
            >
              <label
                class="sc-kfPtPH juLErD"
              >
                <input
                  checked=""
                  hidden=""
                  type="checkbox"
                />
                <span
                  aria-checked="true"
                  aria-label="[든든] 진맛살 1kg 선택"
                  role="checkbox"
                />
              </label>
              <div
                class="sc-jrQzUz ffqCKT"
              >
                <div
                  class="sc-iCfLBT"
                >
                  <div
                    class="sc-furvIG bEiCtt"
                  >
                    <img
                      alt="[든든] 진맛살 1kg"
                      class="sc-pVTma gOPKiY"
                      src="https://cdn-mart.baemin.com/goods/custom/20200525/11157-main-01.png"
                    />
                  </div>
                </div>
                <div
                  class="sc-kDThTU bQjZIX"
                >
                  <h4
                    class="sc-iqsfdx hvMujD"
                  >
                    [든든] 진맛살 1kg
                  </h4>
                  <div
                    class="sc-crHlIS cPCMEP"
                  />
                </div>
                <div
                  class="sc-egiSv ggtrWD"
                >
                  <button
                    aria-label="[든든] 진맛살 1kg 삭제"
                    class="sc-bqiQRQ suzEO"
                    type="button"
                  >
                    <svg>
                      trash-bin.svg
                    </svg>
                  </button>
                  <div
                    class="sc-dkPtyc ibPmRP"
                  >
                    <input
                      aria-label="[든든] 진맛살 1kg 수량 변경 입력"
                      class="sc-hKwCoD kXeTvV"
                      max="99"
                      min="1"
                      type="number"
                      value="1"
                    />
                    <div
                      class="sc-eCImvq isEnoV"
                    >
                      <button
                        aria-label="[든든] 진맛살 1kg 수량 변경 1개 증가"
                        class="sc-jRQAMF crVgkL"
                      >
                        <img
                          alt="증가"
                          class="sc-gKckTs fNcpol"
                          src="arrow-up.svg"
                        />
                      </button>
                      <button
                        class="sc-jRQAMF crVgkL"
                      >
                        <img
                          alt="감소"
                          aria-label="[든든] 진맛살 1kg 수량 변경 1개 감소"
                          class="sc-gKckTs fNcpol"
                          src="arrow-down.svg"
                        />
                      </button>
                    </div>
                  </div>
                  <div
                    aria-label="[든든] 진맛살 1kg 합산 가격"
                  >
                    7,100
                     원
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-fKVsgm bTeRLh"
      >
        <h3
          class="sc-bBHwJV hCzjLe"
        >
          결제예상금액
        </h3>
        <div
          class="sc-iwjezw dYJmqN"
        >
          <div
            class="sc-cxpRKc jWStsh"
          >
            <span
              class="sc-gsDJrp cVrBFO"
              color="#94E0DD"
              font-size="1.25rem"
            >
              결제예상금액
            </span>
            <span
              aria-label="결제예상금액"
              class="sc-gsDJrp cVrBFO"
              color="#94E0DD"
              font-size="1.25rem"
            >
              26,700원
            </span>
          </div>
          <button
            class="sc-bdvvaa jnIuRd"
            color="#FFFFFF"
            font-size="1.5rem"
            height="74px"
            type="button"
            width="100%"
          >
            주문하기(
            2
            개)
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
